### PR TITLE
Add more links to the menu

### PR DESCRIPTION
**Pull request by willdurand**
Refs #2745

---

This PR adds more links to the menu to be able to submit add-ons. I also took the opportunity to clean a few things related to the menu.

screen shot 2017-09-13 at 15 57 13

capture d ecran - 2017-09-13 a 15 57 22

### CODE REVIEW

File: src/amo/components/Link.js [hunk 1]
```diff
@@ -19,7 +19,7 @@ export class LinkBase extends React.Component {
     prependClientApp: PropTypes.bool,
     prependLang: PropTypes.bool,
     to: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),
-    dispatch: PropTypes.func.isRequired,
+    dispatch: PropTypes.func,
```
Comment (kumar303) [inline src/amo/components/Link.js]: What is dispatch used for in this component? I guess it's in the destructor only to make sure it gets excluded from `customProps`?

File: src/ui/components/DropdownMenu/index.js [hunk 2]
```diff
@@ -11,7 +11,8 @@ import './styles.scss';
 
 type Props = {|
   text: string,
-  children: Array<DropdownMenuItem>,
+  // TODO: Replace Array by React.ChildrenArray when we upgrade Flow to 0.53+.
```
Comment (kumar303) [inline src/ui/components/DropdownMenu/index.js]: Ah, yeah, I saw your comment about that. I'll try to look at that upgrade soon. We have some things to fix.

File: tests/unit/amo/components/TestHeader.js [hunk 3]
```diff
@@ -77,17 +76,13 @@ describe(__filename, () => {
...
-      .expects('logOutFromServer')
-      .once()
-      .returns(Promise.resolve());
+    const wrapper = renderHeader({ store, handleLogOut });
 
     const onClick = wrapper.find(DropdownMenuItem).last().prop('onClick');
     onClick(createFakeEvent());
 
-    mockApi.verify();
+    sinon.assert.called(handleLogOut);
```
Comment (kumar303) [inline tests/unit/amo/components/TestHeader.js]: This needs to also make sure `handleLogOut` was called with `{ api }` since that's a crucial part of its signature. This is the type of thing Flow can be really good for. It would tell you if you're calling `handleLogOut()` incorrectly even though it's a stub.

File: tests/unit/core/components/TestAuthenticateButton.js [hunk 4]
```diff
@@ -132,4 +109,33 @@ describe('<AuthenticateButton />', () => {
...
+      const handleLogOut = createHandleLogOutFunction(store.dispatch);
+      const apiState = state.api;
+      // This makes sure the state contains a token because user is logged in.
+      expect(apiState.token).toBeTruthy();
+
+      return handleLogOut({ api: apiState })
+        .then(() => {
+          sinon.assert.calledWith(api.logOutFromServer, { api: apiState });
+          // This makes sure the logOutUser() action has been dispatched.
+          expect(store.getState().api.token).toBeFalsy();
```
Comment (kumar303) [inline tests/unit/core/components/TestAuthenticateButton.js]: Why is this needed? Checking that the API was called correctly is enough, no?

File: src/amo/components/Link.js (hunk 1 above)
Comment (willdurand) [reply src/amo/components/Link.js]: Yes, this is needed here because I want to exclude it from the list passed to `a`.

File: tests/unit/core/components/TestAuthenticateButton.js (hunk 4 above)
Comment (willdurand) [reply tests/unit/core/components/TestAuthenticateButton.js]: That was part of the previous test I merged and the spec reflected it so I thought that was cool to keep.

Comment (willdurand) [reply tests/unit/core/components/TestAuthenticateButton.js]: Ah, edit: this is also because once `logOutFromServer()` has been called, we dispatch the `logOutUser()` action, and the result is that the token should be null'ed.

File: tests/unit/amo/components/TestHeader.js (hunk 3 above)
Comment (willdurand) [reply tests/unit/amo/components/TestHeader.js]: ah sure!

File: src/amo/components/Link.js (hunk 1 above)
Comment (kumar303) [reply src/amo/components/Link.js]: How does `Link` get the dispatch property to begin with? I don't see a mapper for it.

Comment (willdurand) [reply src/amo/components/Link.js]: I think it is because of the spread everywhere 😞

File: tests/unit/core/components/TestAuthenticateButton.js (hunk 4 above)
Comment (kumar303) [reply tests/unit/core/components/TestAuthenticateButton.js]: Okay, I understand. It's not necessary to test that the state gets reduced because that's already covered in https://github.com/mozilla/addons-frontend/blob/071db107cb4fdac16a28071dfa60400164ed501a/tests/unit/core/reducers/test_api.js#L23

In other words, this code should rely on the interface (which is the `logOutFromServer` action), not its implementation.

You can do this by asserting that the action is dispatched:

```js
const dispatchSpy = sinon.spy(store, 'dispatch');
...
sinon.assert.calledWith(dispatchSpy, logOutFromServer());
```

File: src/amo/components/Link.js (hunk 1 above)
Comment (willdurand) [reply src/amo/components/Link.js]: Going to remove actually. It does not fix the real issue.

**Comment by willdurand** [comment]
I have updated the tests!

File: tests/unit/amo/components/TestHeader.js [hunk 5]
```diff
@@ -77,17 +76,15 @@ describe(__filename, () => {
...
-    const mockApi = sinon.mock(api);
+    const handleLogOut = sinon.stub();
 
-    mockApi
-      .expects('logOutFromServer')
-      .once()
-      .returns(Promise.resolve());
+    const wrapper = renderHeader({ store, handleLogOut });
 
     const onClick = wrapper.find(DropdownMenuItem).last().prop('onClick');
```
Comment (kumar303) [inline tests/unit/amo/components/TestHeader.js]: Using `.last()` seems a bit fragile. You could give it a class like `Header-logOut` and find that instead.

File: tests/unit/core/components/TestAuthenticateButton.js [hunk 6]
```diff
@@ -132,4 +111,44 @@ describe('<AuthenticateButton />', () => {
...
+  });
+
+  describe('createHandleLogOutFunction()', () => {
+    it('logs out a signed-in user on the server and client sides', () => {
+      sinon.stub(api, 'logOutFromServer').returns(Promise.resolve());
+
+      const { store, state } = dispatchSignInActions();
+      const dispatchSpy = sinon.spy(store, 'dispatch');
+
+      const handleLogOut = createHandleLogOutFunction(dispatchSpy);
```
Comment (kumar303) [inline tests/unit/core/components/TestAuthenticateButton.js]: Heh, I had to looks this up in the docs. I was going to suggest passing `store.dispatch` since the intention is a bit more clear but they are identical so it's fine either way.

**Comment by muffinresearch** [comment]
FWIW: The earlier test fail was an intermittent failure on just the uitests - cleared by re-starting the job on CI.
